### ISSUE TITLE
Plugins: allow overriding internal plugins, e.g., printer for JS or CSS

### ISSUE THREAD

**Issue by jsnajdr** [prettier/prettier#5919]
**Steps to reproduce:**
Create a plugin that exports a custom printer for JS and/or CSS:
```js
module.exports = {
  printers: {
    estree: myJSPrinter,
    postcss: myCSSPrinter
  }
}
```

Then run Prettier with this plugin on a JS or CSS file.

**Expected behavior:**
The plugin's printer is the one that's used for formatting.

**Actual behavior:**
The default printer is used, no matter what the plugin offers. That's because the array of plugins has the [internal plugins first, the external plugins last](https://github.com/prettier/prettier/blob/aec3978a134a01e62324db51985b7047cf57d71c/src/common/load-plugins.js#L93), and then the printer to use is the [first one that matches the AST format](https://github.com/prettier/prettier/blob/aec3978a134a01e62324db51985b7047cf57d71c/src/main/options.js#L109-L111).

**Why would anyone need it?**
The WordPress community uses a formatting style with a lot of spaces inside parens:
```jsx
function Label( { date } ) {
  const [ translate ] = useTranslate();
  return ( <span>{ translate( 'Updated %s', moment( date ).fromNow() ) }</span> );
}
```

We've been maintaining a [fork that adds a `--paren-spacing` option](https://github.com/automattic/wp-prettier) for the last two years. Now we would like to turn it into a plugin and publish it to NPM as `prettier-plugin-wordpress`.

Then users could install the official `prettier` package and configure it to use the `wordpress` plugin. All editor integrations that [search for](https://github.com/prettier/prettier-atom/blob/8e8be934712882a8902242891e80e3df6f793f6d/src/helpers/getPrettierPath.js#L8) `node_modules/prettier/index.js` and other packages that have `prettier` as peer dependency would use the right version.

Thoughts? Would the team be interested in a PR that swaps the order of the plugin array or searches in it using `findLast`?

**Comment by lydell** [comment]
Maybe you can use npm aliases coming in [npm 6.9.0](https://github.com/npm/cli/releases/tag/v6.9.0-next.0)? I'm thinking you could install your fork under the name "prettier" to "fool" all tools.

**Comment by brodybits** [comment]
@jsnajdr your `--paren-spacing` option is now included in the latest published release of my [`prettierx`](https://www.npmjs.com/package/prettierx) fork, along with a few other recent enhancements:

- fix plugin directory search (PR #5819)
- LWC parser (#5800)
- use [`@typescript-eslint/typescript-estree`](https://www.npmjs.com/package/@typescript-eslint/typescript-estree) version `1.4.1` (they are at `1.4.2` now)

I continue to hope to see this kind of issue resolved to alleviate the need for this kind of a fork.

**Comment by jsnajdr** [comment]
> Maybe you can use npm aliases coming in [npm 6.9.0](https://github.com/npm/cli/releases/tag/v6.9.0-next.0)? I'm thinking you could install your fork under the name "prettier" to "fool" all tools.

Wow, I didn't know that aliases are coming to NPM. And that they already are in Yarn. There used to be some very strong language in npm/npm#2943 saying that it will never happen. Apparently competition can do miracles.

Still, being able to override stuff for the built-in languages could provide a better developer experience. We found that there is some pushback when we advice people to "please use this fork", as opposed to "please use this plugin". The latter is perceived as much more acceptable, although it's technically almost the same thing.

**Comment by lydell** [comment]
I also have to ask – have you considered avoiding the technical issues by changing your code style instead to one supported by Prettier?

**Comment by jsnajdr** [comment]
> I also have to ask – have you considered avoiding the technical issues by changing your code style instead to one supported by Prettier?

Yes, it's been considered in WordPress Core some time ago and the conclusion is summarized in this closing comment: https://github.com/WordPress/gutenberg/issues/2819#issuecomment-361035264

We're trying to provide a tool for the WordPress ecosystem, which is rather big, with many existing independent codebases and established code style conventions. Creating a customized version with one extra option still seems like the most viable path to achieve significant Prettier adoption there.

**Comment by remcohaszing** [comment]
I would like to create a prettier plugin to resolve #4040. I need to override the printer for `package.json` to accomplish this. This is also not possible due to this issue.

**Comment by gziolo** [comment]
> Maybe you can use npm aliases coming in [npm 6.9.0](https://github.com/npm/cli/releases/tag/v6.9.0-next.0)? I'm thinking you could install your fork under the name "prettier" to "fool" all tools.

npm aliases were completely unreliable in npm, and it became only worse since new behavior for peer dependencies was added in npm 7. More details in https://github.com/WordPress/gutenberg/issues/21872. I guess npm aliases are fine for a consuming project as a workaround, but using them in packages distributed on npm creates random errors.
